### PR TITLE
SQ-54/Twitter feed improvements

### PR DESCRIPTION
**Pull request by tiwiz**
We added `Spannable` for mentions and hashtags

|Before|After|
|-------|-----|
|before|after|

And we also are proud to present the first ever _wanna-be-infinite-recyclerview-but-they-dont-so-its-limited-were-sorry_

|Before|After|
|------|-----|
|before|after|

Known issues:
- Twitter returns us the list of the URLs included in the single tweet, so that we can simply use this list to create our `LinkableSpannable`. Problem is sometimes they lose some of the URLs, so we might want to do it manually to ensure we don't lose anything
- We don't manage images or videos yet

### CODE REVIEW

File: app/src/main/java/net/squanchy/tweets/TweetsPageView.java [hunk 1]
```diff
@@ -76,6 +79,8 @@ private void initList() {
 
         if (!isInEditMode()) {
             SearchTimeline timeline = new SearchTimeline.Builder()
+                    .resultType(SearchTimeline.ResultType.RECENT)
+                    .maxItemsPerRequest(10)
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: Shall we do more? Like, 50?

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import net.squanchy.tweets.view.TimelineStateHolder;
+
+public class TwitterService<T extends Identifiable> {
+
+    private static final long CAPACITY = 200L;
+    private final Timeline<T> timeline;
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: Whiteline after the constant plz

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    private final List<T> itemList;
+
+    public TwitterService(Timeline<T> timeline) {
+        this.timeline = timeline;
+        this.timelineStateHolder = new TimelineStateHolder();
+
+        itemList = new ArrayList<>();
+    }
+
+    public void refresh(Callback<TimelineResult<T>> developerCb) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: What is the meaning behind `developerCb`? Why "developer callback"?

Besides I'd rather hide Twitter's callbacks. You should have one "get data" method that returns `Observable<List<Tweet>>` that you keep appending items to, backed by a subject, and the `refresh` method shouldn't have any callback. We don't really care if it worked or not.

I suspect (haven't gotten there yet) that you're using the callback to determine whether to show the loader at the bottom, but why not always showing it? If you get to the bottom (or close to it) a request is fired anyway. It'll always be "loading" at the bottom. If you really want to cap the number of items (see below) then the visibility of that can simply be based on `items.size() < MAX_ITEMS)_COUNT`.

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import com.twitter.sdk.android.core.models.Identifiable;
+import com.twitter.sdk.android.tweetui.Timeline;
+import com.twitter.sdk.android.tweetui.TimelineResult;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import net.squanchy.tweets.view.TimelineStateHolder;
+
+public class TwitterService<T extends Identifiable> {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: Can we please order the methods in this class to be top-down? It's very hard to follow the logic  this way. Also, what is `Identifiable`?

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                timeline.next(minPosition, cb);
+            } else {
+                cb.failure(new TwitterException("Request already in flight"));
+            }
+        } else {
+            cb.failure(new TwitterException("Max capacity reached"));
+        }
+    }
+
+    private boolean withinMaxCapacity() {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: I don't understand why we need this?

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+        RefreshCallback(Callback<TimelineResult<T>> developerCb, TimelineStateHolder timelineStateHolder) {
+            super(developerCb, timelineStateHolder);
+        }
+
+        @Override
+        public void success(Result<TimelineResult<T>> result) {
+            if (!result.data.items.isEmpty()) {
+                itemList.clear();
+                itemList.addAll(result.data.items);
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: I'd rather swap the list tbh

File: app/src/main/java/net/squanchy/tweets/view/ScrollListener.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
+package net.squanchy.tweets.view;
+
+import android.support.v7.widget.LinearLayoutManager;
+import android.support.v7.widget.RecyclerView;
+
+public abstract class ScrollListener extends RecyclerView.OnScrollListener {
+
+    private static final int THRESHOLD = 2;
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/view/ScrollListener.java]: White line  after the constants

File: app/src/main/java/net/squanchy/tweets/view/ScrollListener.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    private final LinearLayoutManager layoutManager;
+
+    protected ScrollListener(LinearLayoutManager layoutManager) {
+        this.layoutManager = layoutManager;
+    }
+
+    @Override
+    public void onScrolled(RecyclerView view, int dx, int dy) {
+
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/view/ScrollListener.java]: Remove this empty line please

File: app/src/main/java/net/squanchy/tweets/view/TimelineStateHolder.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+    public boolean startTimelineRequest() {
+        return requestInFlight.compareAndSet(false, true);
+    }
+
+    public void finishTimelineRequest() {
+        requestInFlight.set(false);
+    }
+}
+
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/view/TimelineStateHolder.java]: Extra whiteline, remove plz

File: app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java [hunk 10]
```diff
@@ -8,42 +8,63 @@
...
+        this.repository = new TwitterRepository(timeline);
+    }
+
+    @Override
+    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, @TweetViewTypeId int viewType) {
+        if (viewType == TweetViewTypeId.TWEET) {
+            View view = LayoutInflater.from(context).inflate(R.layout.item_tweet, parent, false);
+            return new TweetViewHolder(view);
+        } else if (viewType == TweetViewTypeId.LOADING) {
+            Timber.d("Creating loading view");
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java]: Can we remove the debug logs from here and elsewhere please?

File: app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java [hunk 11]
```diff
@@ -8,42 +8,63 @@
...
+        return repository.getItemViewType(position);
     }
 
     @Override
-    public void onBindViewHolder(TweetViewHolder holder, int position) {
-        holder.updateWith(delegate.getItem(position));
+    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
+        if (repository.getItemViewType(position) == TweetViewTypeId.TWEET) {
+            ((TweetViewHolder) holder).updateWith(repository.itemAt(position));
+        }
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java]: Missing the `else if() /* do nothing */` and the `else throw`

**Review by rock3r — CHANGES_REQUESTED** [review]

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java (hunk 4 above)
Comment (tiwiz) [reply app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: Identifiable is from Twitter, it's `Tweet` superclass

File: app/src/main/java/net/squanchy/tweets/view/ScrollListener.java (hunk 7 above)
Comment (tiwiz) [reply app/src/main/java/net/squanchy/tweets/view/ScrollListener.java]: Done and committed, but it looks like GitHub doesn't like it :-\

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java (hunk 4 above)
Comment (tiwiz) [reply app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: We can actually remove the generic and just go with `Tweet` as we won't be using it with anything else

File: app/src/main/java/net/squanchy/tweets/TweetsPageView.java (hunk 1 above)
Comment (quiro91) [reply app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: We went for 10 because our viewholder is quite big and there's already some scrolling to do before reaching the end. I'd say we can go for 20 or 30 (which is the default value)

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java (hunk 3 above)
Comment (tiwiz) [reply app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: @rock3r actually we already always show the loader at the bottom, the callback is just the way Twitter does the things, but I agree on wrapping it with RxJava

File: app/src/main/java/net/squanchy/tweets/view/ScrollListener.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    private final LinearLayoutManager layoutManager;
+
+    protected ScrollListener(LinearLayoutManager layoutManager) {
+        this.layoutManager = layoutManager;
+    }
+
+    @Override
+    public void onScrolled(RecyclerView view, int dx, int dy) {
+        int totalItemCount = layoutManager.getItemCount();
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/tweets/view/ScrollListener.java]: do you need the layout manager as a dependency? You receive the recycler view as a parameter

File: app/src/main/java/net/squanchy/tweets/TweetsPageView.java [hunk 13]
```diff
@@ -84,12 +89,26 @@ private void initList() {
         }
 
         emptyView.setText(context.getString(R.string.no_tweets_for_query, query));
+
+        LinearLayoutManager layoutManager = (LinearLayoutManager) tweetsList.getLayoutManager();
+
+        scrollListener = new ScrollListener(layoutManager) {
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: Can we have the scroll listener accept a `LoadMoreListener` which has only that method instead of extending the class itself?

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        return itemList.size();
+    }
+
+    public Tweet itemAt(int position) {
+        return itemList.get(position);
+    }
+
+    private void load(Long minPosition, Callback<TimelineResult<Tweet>> cb) {
+        if (timelineStateHolder.startTimelineRequest()) {
+            timeline.next(minPosition, cb);
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: what happens if `minPosition` is null here? (which can be in case `timelineStateHolder` returns a null there)

File: app/src/main/java/net/squanchy/tweets/view/TimelineStateHolder.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
+package net.squanchy.tweets.view;
+
+import com.twitter.sdk.android.tweetui.TimelineCursor;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class TimelineStateHolder {
+
+    private TimelineCursor nextCursor;
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/tweets/view/TimelineStateHolder.java]: can we make both of the `TimelineCursor`s optional? so that if someone deals with this code will know they have to check for presence

File: app/src/main/java/net/squanchy/tweets/view/TimelineStateHolder.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    private TimelineCursor previousCursor;
+    private final AtomicBoolean requestInFlight = new AtomicBoolean(false);
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/tweets/view/TimelineStateHolder.java]: do you use this? I think you never read the value of this

File: app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java [hunk 17]
```diff
@@ -8,42 +8,66 @@
...
+public class TweetsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
 
     private final Context context;
-    private final TimelineDelegate<Tweet> delegate;
+    private final TwitterRepository repository;
 
     public TweetsAdapter(Timeline<Tweet> timeline, Context context) {
         this.context = context;
-        this.delegate = new TimelineDelegate<>(timeline);
+        this.repository = new TwitterRepository(timeline);
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java]: can you have the repository as a dependency here instead of the timeline? :ghost: 

On another note: I don't think the repository should be inside the adapter, the adapter should just _receive_ data (i.e. with an `updateWith(xxx)` method) and adapt it for the view, the repository should be used and then the results fed into the adapter IMHO.

File: app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java [hunk 18]
```diff
@@ -8,42 +8,66 @@
...
 
-public class TweetsAdapter extends RecyclerView.Adapter<TweetViewHolder> {
+public class TweetsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
 
     private final Context context;
-    private final TimelineDelegate<Tweet> delegate;
+    private final TwitterRepository repository;
 
     public TweetsAdapter(Timeline<Tweet> timeline, Context context) {
         this.context = context;
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java]: I think you can safely have the `LayoutInflater` instead of the context as a field since the context is used only for that.

**Review by fourlastor — DISMISSED** [review]
Nice! I think we need a bit more separation between the view and the service but that's a whole bunch of good stuff :smile:

**Comment by rock3r** [comment]
Putting :x: label while @tiwiz and @quiro91 rework this

File: app/src/main/java/net/squanchy/tweets/TweetsPageView.java [hunk 19]
```diff
@@ -1,30 +1,34 @@
...
+import static net.squanchy.support.ContextUnwrapper.unwrapToActivityContext;
+
 public class TweetsPageView extends LinearLayout {
 
     private TextView emptyView;
     private RecyclerView tweetsList;
     private TweetsAdapter tweetsAdapter;
     private SwipeRefreshLayout swipeLayout;
+    private TwitterService twitterService;
+    private Disposable disposable;
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: Can we call this `subscription` like in the rest of the app?

File: app/src/main/java/net/squanchy/tweets/TweetsPageView.java [hunk 20]
```diff
@@ -69,32 +72,42 @@ protected void onAttachedToWindow() {
         initList();
     }
 
+    @Override
+    protected void onDetachedFromWindow() {
+        super.onDetachedFromWindow();
+        if (disposable != null && !disposable.isDisposed()) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: No need to do those checks (`dispose()` is idempotent, and if you get here without having ever subscribed, there is a bug and I'd rather crash early in that case

File: app/src/main/java/net/squanchy/tweets/domain/view/HashtagEntity.java [hunk 21]
```diff
@@ -0,0 +1,17 @@
+package net.squanchy.tweets.domain.view;
+
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class HashtagEntity {
+
+    public static HashtagEntity create(String text, int start, int end){
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/domain/view/HashtagEntity.java]: cmd-alt-l

File: app/src/main/java/net/squanchy/tweets/domain/view/MentionEntity.java [hunk 22]
```diff
@@ -0,0 +1,17 @@
+package net.squanchy.tweets.domain.view;
+
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class MentionEntity {
+
+    public static MentionEntity create(String displayName, int start, int end){
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/domain/view/MentionEntity.java]: cmd-alt-l

File: app/src/main/java/net/squanchy/tweets/domain/view/UrlEntity.java [hunk 23]
```diff
@@ -0,0 +1,17 @@
+package net.squanchy.tweets.domain.view;
+
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class UrlEntity {
+
+    public static UrlEntity create(String url, int start, int end){
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/domain/view/UrlEntity.java]: cmd-alt-l

File: app/src/main/java/net/squanchy/tweets/service/TwitterRepository.java [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+import retrofit2.Call;
+
+public class TwitterRepository {
+
+    private static final int MAX_ITEM_PER_REQUEST = 100;
+
+    private final SearchService searchService;
+    private final String query;
+
+    public TwitterRepository(String query) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/service/TwitterRepository.java]: Can we pass in the query to `load()` instead? So we don't have to pass it aaaall the way through Dagger which is a bit weird

File: app/src/main/java/net/squanchy/tweets/util/TwitterDateFormatter.java [hunk 25]
```diff
@@ -23,11 +22,11 @@ private TwitterDateFormatter() {
 
     public static String getTimestampFrom(Tweet tweet, Context context) {
 
-        if (tweet.createdAt == null) {
+        if (tweet.createdAt() == null) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/util/TwitterDateFormatter.java]: I think this can never happen, AutoValue would blow up if this were null as it's not marked as `@Nullable`—which is great, _if_ it is actually not mandatory I'd rather have it `Optional` than `@Nullable`

File: app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java [hunk 26]
```diff
@@ -31,19 +30,27 @@ public TweetViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
...
 
     public boolean isEmpty() {
-        return delegate.getCount() == 0;
+        return tweetList.isEmpty();
     }
 
-    public void refresh(TweetsPageView.TimelineLoadingCallback timelineLoadingCallback) {
-        delegate.refresh(timelineLoadingCallback);
+    public void updateWith(List<Tweet> tweets) {
+        if (tweets.isEmpty()) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java]: This is not correct ⚠️  You should always replace the tweets list with the new one (make it not final and swap them, rather than mutating), otherwise we wouldn't be able to show the empty state for example when needed.

File: app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java [hunk 27]
```diff
@@ -31,19 +30,27 @@ public TweetViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
...
+            return;
+        }
+
+        tweetList.clear();
+        final ArrayList<Tweet> receivedItems = new ArrayList<>(tweets);
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/view/TweetsAdapter.java]: No need for `final` here

**Review by rock3r — CHANGES_REQUESTED** [review]
ALMOST THERE! This is perfectly good for our MVP, we'll make it fancier later :) A few minor comments and a probably incorrect `if`, then good to go

File: app/src/main/java/net/squanchy/tweets/TweetsPageView.java [hunk 28]
```diff
@@ -69,32 +73,40 @@ protected void onAttachedToWindow() {
         initList();
     }
 
+    @Override
+    protected void onDetachedFromWindow() {
+        super.onDetachedFromWindow();
+        subscription.dispose();
+    }
+
     private void initList() {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: This should be before `onDetachedFromWindow`, but... can we move:
* the injection out of here and into the ctor
* the adapter creation in `onFinishInflate()`
* and inline everything else in `onAttachedToWindow()`?

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+                .id(tweet.id)
+                .text(tweet.text)
+                .createdAt(tweet.createdAt)
+                .hashtags(mapHashtags(tweet.entities.hashtags))
+                .mentions(mapMentions(tweet.entities.userMentions))
+                .urls(mapUrls(tweet.entities.urls))
+                .build();
+    }
+
+    private List<HashtagEntity> mapHashtags(List<com.twitter.sdk.android.core.models.HashtagEntity> entities) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/service/TwitterService.java]: Can we rename these `map` methods to `parse`

File: app/src/main/java/net/squanchy/tweets/TweetsPageView.java (hunk 28 above)
Comment (quiro91) [reply app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: I may be wrong but we need the `Activity` context for the injection, should we really do it in the ctor? In the other views we do it in `onFinishInflate()`. Also, shall we implement `LifecycleView` and use its callbacks instead of `onAttachedToWindow()` and `onDetachedFromWindow()` ?

Comment (rock3r) [reply app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: You have the (activity) context in the ctor as a parameter... A view cannot be attached to an activity that didn't create it. Also yes please implement the interface, I had _absolutely_ forgotten about its existence 💯

Comment (rock3r) [reply app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: @fourlastor says "as long as you remember to call the methods because they don't get magically called". I suppose it means see how the other `LifecycleViews` are handled.

Comment (quiro91) [reply app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: @fourlastor that's weird. From what I can see `LifecycleViews` are stored in an `ArrayList` in the `HomeActivity` which forwards those callback to each views in a for loop. So I'd just need to add the view to the list

Comment (rock3r) [reply app/src/main/java/net/squanchy/tweets/TweetsPageView.java]: Yeah I think that's what he meant

**Review by rock3r — APPROVED** [review]
🎉
